### DISCUSSION TITLE
can you add support for mebay generator controller

### DISCUSSION THREAD

**Discussion by yellowdre** [jgyates/genmon#1159] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

i have a mebay generator controller and i have the modbus  registers but i dont know how to get it working with genmon
can someone please help me with this, i have include modbus manual
[DC6xD MODBUS_Communication protocol_V1.2-20231124.pdf](https://github.com/user-attachments/files/16413771/DC6xD.MODBUS_Communication.protocol_V1.2-20231124.pdf)

**Comment by jgyates** [comment]
Hi @yellowdre 

[This wiki article](https://github.com/jgyates/genmon/wiki/Appendix-N-Genmon-Supporting-Other-Controller-Types#the-process-of-creating-a-new-controller-support-within-genmon) has the process for adding support for new generator controller.

There are 4 steps described in the above link. If have provide number 1, if you can provide number 2, I can add support my completing number 3.

In short you need to connect a pi to your controller via the serial interface. I assume this is RS-485? is that correct? Once you do that you can use the modbusdump.py program to read the registers. I will use the register data provided via modbusdump.py along with the data in the attached PDF above to complete step 3.

Let me know if you have questions on how to proceed.

**Reply by jgyates** [reply, marked as answer]
I have added support for mebay DC4x - DC9x controllers. I have only tested with registers from a DC42 so let me know how testing goes. any feedback is welcome. Submitting your registers and logs from the about page helps me track down issues so if you do try this please send the feedback.

I move the issue thread [here](https://github.com/jgyates/genmon/issues/1216) to this thread.

The wiki page has been updated. 

@maindrive
if you are testing this let me know how it goes. Also, if you have a bulleted list of any steps you are taking to set this up on the generator controller side please send them to me and I can create a wiki page for the MEBAY controller. pics and / or screen shots are great.

Also any info on what things you did on the genmon side (e.g. RS-485 adapter, cable, etc) would help also.

I can make the wiki if you just give me the data.

Let me know how your testing goes.
@jgyates

**Reply by jgyates** [reply]
I also renamed the json file that you sent in the other thread so you will need to update the software from the About page, then set the controller on the advanced settings. Also, the frequency, RPM, rated kw, will be set on the Settings page, not in the JSON file so there is no need to edit that.

**Reply by jgyates** [reply]
I contacted Mebay and they sent over modbus docs for more controllers so I will look at this tomorrow to see how many can be supported.

**Reply by maindrive** [reply]
Woo Hoo!, That's great news..
Please change the E Stop from 1 and change to 0. 
The current time (Holding:1008) keeps flashing red.
Yes, I will send a write up of what all I have done.
The biggest hang up was the CRC check setting, it must be changed to  ***** CRC:L-H ***** in the RS485 Port  settings in the controller.
My Controller Default is CRC:H-L 

Again Big Thanks For Your Support!

**Reply by jgyates** [reply]
on the registers live view page in the genmon web interface, if a register changes it turns red and then it will become a lighter red over time. If the register changes again then it will go back to the darker red. This happening for registers that change, like the time, is expected.

I checked in an update that changes the polarity of the emergency stop bit.

Let me know what else you find.

**Comment by maindrive** [comment]
Hello Jason,

I have included my MEBAY_DC42D_MK3_With_Commands.txt, the command are 
working, but the Engine State : Alarms do not print out under the Engine 
State.
Engine State statuses work fine, I would expect when in alarm the alarm 
statuses would be printed like: Crank failure, Emergency stop, RPM 
Signal lost, some where on the page?

I have tested:

"2": "Under speed",
"19": "RPM Signal lost",
"46": "Emergency stop",
"47": "Crank failure",

//

Here is what I have tested.
Please find attached zip file with pics

The main stumbling block need to set 485 CRC Setting to CRC:L_H in the 
rs485 communication port setting in the controller

Tested with the following:
#Pi Zero 2 setup
https://www.waveshare.com/rs485-can-hat.htm
RS485 CAN HAT for Raspberry Pi
RS485 CAN HAT
#Add driver
sudo nano /boot/config.txt or /boot/firmware/config.txt
sudo nano /boot/firmware/config.txt
dtparam=spi=on
dtoverlay=mcp2515-can0,oscillator=12000000,interrupt=25,spimaxfrequency=2000000

This creates /dev/serial0 -> /dev/ttyAMA0
commands used to test
sudo ./genenv/bin/python3 ./OtherApps/modbusdump.py -r 19200 -p 
/dev/ttyAMA0 -a 10 -s 4096 -e 4191
sudo ./genenv/bin/python3 ./OtherApps/modbusdump.py -r 19200 -p 
/dev/serial0 -a 10 -s 4096 -e 4191


#Raspberry Pi 4 setup
https://www.waveshare.com/rs485-rs232-hat.htm
Isolated RS485 RS232 Expansion HAT for Raspberry Pi, SPI Control
RS485 RS232 HAT
#Add driver
sudo nano /boot/config.txt or /boot/firmware/config.txt
sudo nano /boot/firmware/config.txt
dtoverlay=sc16is752-spi1,int_pin=24

This creates /dev/ttySC0 rs485 port and /dev/ttySC1 RS232 port
command used to test
sudo ./genenv/bin/python3 ./OtherApps/modbusdump.py -r 19200 -p 
/dev/ttySC0 -a 10 -s 4096 -e 4191


# for Laptop or Raspberry Pi
DTECH USB to RS422 RS485 Serial Port Converter Adapter Cable
port settings Baud: 19200, Data bits: 8, Party: None, Stop bits: 1, Flow 
control: none

Raspberry pi
command used to test
sudo ./genenv/bin/python3 ./OtherApps/modbusdump.py -r 19200 -p 
/dev/ttyUSB0 -a 10 -s 4096 -e 4191


# Network
Note: The PUSR USR-W610 and Waveshare RS232/485 TO WIFI ETH setup web 
pages look the same!
See screen shots

https://www.pusr.com/products/rs232-rs485-to-wifi-converters-usr-w610.html
PUSR Serial to WiFi Ethernet Wireless Converter RS232 RS485 Serial 
Server(USR-W610)

https://www.waveshare.com/product/iot-communication/wired-comm-converter/ethernet-to-uart-rs232-rs485/rs232-485-to-wifi-eth-b.htm?sku=25222
Waveshare Industrial Grade Serial Server RS232/RS485 to WiFi and 
Ethernet
SKU:25222, Part No.:RS232/485 TO WIFI ETH (B)

command used to test
sudo ./genenv/bin/python3 ./OtherApps/modbusdump.py -i 192.168.1.143 -t 
8899 -a 10 -s 4096 -e 4195

**Reply by jgyates** [reply]
Thanks for the update. It looks like you attachement was removed by github. If you respond to the thread in email, github will remove the attachment. You can either post the attachment via the github web interface or sent it to my email address. I will incorporate these changes into the projects and make a wiki page for the mebay controller with the data you provided.

The active alarms and warnings are suppose to print in the status page under the heading active alarms. This is done via the "alarm_status" and "alarm_conditions" objects in the JSON file. If alarm_status is true, then it will display alarm_conditions. You should also get a notification via email or add on (email gives more info) if an alarm is active.

Also, if you have email setup and fucntional can you submit your log files and registers on the about page? Submit them on an active alarm if possible.

**Reply by jgyates** [reply]
I have updated the repository with the latest JSON file and I created a wiki entry with your screenshots and data. Thanks!

https://github.com/jgyates/genmon/wiki/Appendix-R---MEBAY-Controller-with-RS-485

**Reply by jgyates** [reply]
When you get a moment it would be helpful to submit your log files on the About page. Outbound email must be enabled for this to work.

**Reply by jgyates** [reply]
I got your logs. It looks like you are still using the old file JSON name for the custom controller 

This should be changed on the advanced settings page by selecting MEBAY_DCxx.json and then deleting the file  file /home/pi/genmon/data/controller/MEBAY_DC42D_MK3_Comands.json

If you do not make this change then future updates will not take effect until you edit your old json file. If you use the newly named file then you will get updates just by updating from the about page.

Other than that there are no unusual things in the log.

**Comment by maindrive** [comment]
In MEBAY_DCxx.json

At line 401
          "reg": "1024", should be "1023" for Average Power factor

**Comment by maindrive** [comment]
At line 620
          "reg": "1034", should be "102BH" Backup Battery voltage

**Comment by maindrive** [comment]
I have the Rated Frequency and Nominal RPM in the settings page, but 
there not showing up on the Maintenance page.

Generator Model Specific Settings

  	Rated Frequency: 60
  	Nominal RPM: 3600
  	Maximum kW Output: 13
  	Fuel Type: Propane


Maintenance :
     Model : MEBAY DCxx Controller
     Maintenance Due : No
     Controller Detected : MEBAY DC-42 Controller
     Nominal RPM : Unknown
     Rated kW : 13
     Nominal Frequency : Unknown
     Phase : 1
     Fuel Type : Propane

**Reply by jgyates** [reply]
I made the corrections you mentioned. Let me know if there are any issues. Also, you should see a setting for Nominal Battery Voltage on the settings page so make sure that is correct (12 or 24 V). This sets the max and scale of the battery voltage gauge on the main web interface page.

There were some issues with the default values if the JSON file specified "Unknown" that were causing the problem you mentioned. If the JSON file specifies "Unknown" then it should use the genmon settings and since the Mebay custom controller JSON could be used on any generator I have made it default to the settings of genmon which are user configurable.

Let me know if you have any other issues.
